### PR TITLE
tests for check_password function

### PR DESCRIPTION
**Pull request by Greta-mart**

### CODE REVIEW

File: tests/threc/hw_password/test_password.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    pytest.param('qwerty@@', id='Password without capital letter and number, but with special character'),
+    pytest.param('Qwerty88', id='Password without special characted, but with capital letter and number'),
+    pytest.param('Qwerty@@', id='Password without number, but with capital letter and special symbol'),
+    pytest.param('Qwe8@', id='Password with capital letter, special symbol, number, but less than 8 symbols'),
+    pytest.param('Qwertyuiopasdfghjklzxcvbnmqw8@', id='Password with more that 30 symbols'),
+    pytest.param("string with 123@", id='Password with several words')
+])
+def test_password_negative(res):
+    assert check_password(res) is False
+
```
Comment (fkudinov-mn) [inline tests/threc/hw_password/test_password.py:32]: 2 empty lines between objects in module

File: tests/threc/hw_password/test_password.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+@pytest.mark.negative('Password with one symbol')
+@pytest.mark.parametrize(["res"], [
+    pytest.param('1', id='Password with one symbol'),
+    pytest.param(' ', id='Password with space'),
+    pytest.param('', id='Empty password'),
+    pytest.param('Q8@e', id='Password with less then 8 symbols'),
+    pytest.param('qwertyuiop', id='Password with small letters'),
+    pytest.param('12345678', id='Password with only numbers'),
+    pytest.param('qwerty8@', id='Password without capital letter, but with nuber and special character'),
```
Comment (fkudinov-mn) [inline tests/threc/hw_password/test_password.py:21]: try to reduce id names, its hard to read

e.g.:

no upper letters
no numbers
too short
too long

File: tests/threc/hw_password/test_password.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+])
+def test_password_positive(res):
+    assert check_password(res)
+
+@pytest.mark.negative('Password with one symbol')
+@pytest.mark.parametrize(["res"], [
+    pytest.param('1', id='Password with one symbol'),
+    pytest.param(' ', id='Password with space'),
+    pytest.param('', id='Empty password'),
+    pytest.param('Q8@e', id='Password with less then 8 symbols'),
```
Comment (fkudinov-mn) [inline tests/threc/hw_password/test_password.py:18]: try to test boundary cases.   

7 symbols, 8, 30, 31

**Review by kudinov-fedor — APPROVED** [review]

File: tests/threc/hw_password/test_password.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+def test_password_negative(res):
+    assert check_password(res) is False
+
+@pytest.mark.checkError
+@pytest.mark.parametrize(['res'], [
+    pytest.param(123, id='Password with integer value instead of string')
+])
+def test_raise_error(res):
+    with pytest.raises(TypeError):
+        assert check_password(res) is False
```
[the remark this answers is not in the document]
Comment (Greta-mart) [reply tests/threc/hw_password/test_password.py]: removed parametrize

File: tests/threc/hw_password/test_password.py (hunk 1 above)
Comment (Greta-mart) [reply tests/threc/hw_password/test_password.py:34]: Added two lines

Comment (fkudinov-mn) [reply tests/threc/hw_password/test_password.py:34]: not here, there should not be empty lines between function and decorator
use flake8 it can show warnings

File: tests/threc/hw_password/test_password.py [hunk 5]
```diff
@@ -10,21 +10,21 @@
 def test_password_positive(res):
     assert check_password(res)
 
-@pytest.mark.negative('Password with one symbol')
+@pytest.mark.negative()
```
Comment (fkudinov-mn) [inline tests/threc/hw_password/test_password.py:13]: call is not required

@pytest.mark.negative

File: tests/threc/hw_password/test_password.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import pytest
+from tests.threc.hw_password.check_password import check_password
+
```
Comment (fkudinov-mn) [inline tests/threc/hw_password/test_password.py:3]: 2 empty lines between imports and function

File: tests/threc/hw_password/test_password.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+@pytest.mark.positive
+@pytest.mark.parametrize(["res"], [
+    pytest.param('Qwerty8@', id='Password with 8 symbols'),
+    pytest.param('Qwertyuiopasdfghjklzxcvbnm12@', id='Password with 30 symbols'),
+    pytest.param('Qwertyuiopasdf12@', id='Password with symbols in between 8 and 30')
+])
+def test_password_positive(res):
+    assert check_password(res)
+
```
Comment (fkudinov-mn) [inline tests/threc/hw_password/test_password.py:12]: 2 empty lines between functions

File: tests/threc/hw_password/test_password.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    pytest.param('qwerty@@', id='Password with lower letters, special character'),
+    pytest.param('Qwerty88', id='Password capital letter and number'),
+    pytest.param('Qwerty@@', id='Password with capital letter and special symbol'),
+    pytest.param('Qwe8@', id='Length for password is too short'),
+    pytest.param('Qwertyuiopasdfghjklzxcvbnmqw8@', id='Too long password'),
+    pytest.param("string with 123@", id='Password with several words')
+])
+
+
+def test_password_negative(res):
```
Comment (fkudinov-mn) [inline tests/threc/hw_password/test_password.py:32]: no empty line between decorator and function

File: tests/threc/hw_password/test_password.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    assert check_password(res) is False
+
+
+# Password with integer value instead of string
```
Comment (fkudinov-mn) [inline tests/threc/hw_password/test_password.py:36]: move comment to docstring of a function

**Review by fkudinov-mn — CHANGES_REQUESTED** [review]
codestyle should be fixed

**Review by kudinov-fedor — CHANGES_REQUESTED** [review]
-

File: tests/threc/not_reliable/not_reliable_function.py [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+    res = random.random()
+    if res < 0.5:
+        raise RuntimeError("Res is less than 0.5:  {}".format(res))
+    return res
+
+
+def test_res_more_0_5(mocker):
+    mocker.patch('random.random', return_value=0.6)
+    res = not_reliable()
+    assert res == 0.6
```
Comment (kudinov-fedor) [inline tests/threc/not_reliable/not_reliable_function.py:18]: function with tests should start with 'test_' prefix, so that pytest could gather these tests

File: tests/threc/not_reliable/test_not_reliable_function.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+
+def test_not_reliable_0_6(mocker):
+    mocker.patch('random.random', return_value=0.6)
+    assert not_reliable() == 0.6
+
+
+def test_not_reliable_2(mocker):
+    mocker.patch('random.random', return_value=2)
+    assert not_reliable() == 2
```
Comment (kudinov-fedor) [inline tests/threc/not_reliable/test_not_reliable_function.py:17]: you can group these 3 tests into parametrize

File: tests/threc/not_reliable/test_not_reliable_function_2.py [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+
+def test_not_reliable_2_1(mocker):
+    mocker.patch.object(result, 'random', return_value=1)
+    assert not_reliable() == 1
+
+
+def test_not_reliable_2_0_6(mocker):
+    mocker.patch.object(result, 'random', return_value=0.6)
+    assert not_reliable() == 0.6
```
Comment (kudinov-fedor) [inline tests/threc/not_reliable/test_not_reliable_function_2.py:25]: you can group positive tests into parametrize

**Review by kudinov-fedor — APPROVED** [review]

File: tests/threc/not_reliable/not_reliable_function.py [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+
+def not_reliable():
+    res = random.random()
+    if res < 0.5:
+        raise RuntimeError("Res is less than 0.5:  {}".format(res))
+    return res
+
+
+def test_res_more_0_5(mocker):
+    mocker.patch('random.random', return_value=0.6)
```
Comment (kudinov-fedor) [inline tests/threc/not_reliable/not_reliable_function.py:16]: you should not use patch like that !!!
patch is from UNITTEST and  pytest does not revert these changes,
better to check it , weather this patch is still applied for next tests


you should use either mocker.patch.object

OR

patch as context manager

```
def test_res_more_0_5(mocker):
    with mocker.patch('random.random', return_value=0.6): 
        res = not_reliable()
        assert res == 0.6
```

**Review by kudinov-fedor — CHANGES_REQUESTED** [review]
retest mocker.patch

File: tests/threc/not_reliable/test_not_reliable_function.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
+from tests.threc.not_reliable.not_reliable_function import not_reliable
+import pytest
+
+
+def test_not_reliable_0_5(mocker):
+    mocker.patch('random.random', return_value=0.5)
```
Comment (kudinov-fedor) [inline tests/threc/not_reliable/test_not_reliable_function.py:6]: you can mock as this, module is a singletone in python - single object with global access

import random
mocker.patch.object(random, 'random', return_value=0.5)

File: tests/threc/not_reliable/test_not_reliable_function.py [hunk 15]
```diff
@@ -2,22 +2,19 @@
...
-    mocker.patch('random.random', return_value=2)
-    assert not_reliable() == 2
-
-
-def test_not_reliable_0(mocker):
+@pytest.mark.parametrize(["value"], [
+    pytest.param(0.5),
+    pytest.param(0.6),
+    pytest.param(2)
+])
```
Comment (kudinov-fedor) [inline tests/threc/not_reliable/test_not_reliable_function.py:9]: @pytest.mark.parametrize("value", [0.5, 0.6, 2])

**Review by kudinov-fedor — APPROVED** [review]
